File: Core/Specifications/OrderSpecification.cs
using Core.Entities.OrderAggregate;
using Microsoft.EntityFrameworkCore.ChangeTracking;
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;
using System.Threading.Tasks;

namespace Core.Specifications
{
    public class OrderSpecification : BaseSpecification<Order>
    {
        public OrderSpecification(string email): base(x => x.BuyerEmail == email)
        {
            AddInclude(x => x.DeliveryMethod);
            AddInclude(x => x.OrderItems);
            AddOrderByDescending(x => x.OrderDate);
        }
        public OrderSpecification(string email,int id) : base(x => x.BuyerEmail == email && x.Id == id)
        {
            AddInclude("DeliveryMethod");
            AddInclude("OrderItems");
            AddOrderByDescending(x => x.OrderDate);
        }
    }
}

File: java-api/core/src/main/java/com/aayuskinet/core/specifications/OrderSpecification.java
package com.aayuskinet.core.specifications;

import com.aayuskinet.core.entities.orderaggregate.Order;
import org.springframework.data.jpa.domain.Specification;

public class OrderSpecification extends BaseSpecification<Order> {

    public OrderSpecification(String email) {
        super(createSpecification(email, null));
        addIncludes();
        addOrderByDescending("orderDate");
    }

    public OrderSpecification(String email, int id) {
        super(createSpecification(email, id));
        addIncludes();
        addOrderByDescending("orderDate");
    }

    private void addIncludes() {
        addInclude("deliveryMethod");
        addInclude("orderItems");
    }

    private static Specification<Order> createSpecification(String email, Integer id) {
        return (root, query, criteriaBuilder) -> {
            var predicate = criteriaBuilder.equal(root.get("buyerEmail"), email);

            if (id != null) {
                predicate = criteriaBuilder.and(predicate, criteriaBuilder.equal(root.get("id"), id));
            }

            return predicate;
        };
    }
}
